### ISSUE TITLE
100.000 X covered region included for Freebayes variant calling?

### ISSUE THREAD

**Issue by NeillGibson** [chapmanb/bcbio-nextgen#1281]
Hi, 

Because of a not finishing Freebayes analysis I noticed that a 200 bp region with 100.000 X coverage was included variant calling. 

The total callable region BED record was 3.000 bp with also more coverage peaks going up to 5.000 X but the majority of the region was the expected 30 X. 

Could it be that there is a bug in the BCBio part that calculates based on coverage in the BED files which regions are suitable for variant calling?

I work with a lot of different reference genomes, some custom and or in-house and with varying degrees of assembly quality. Manually creating blacklist for each reference to exclude collapsed repeat regions and centromeres is too much work. 

Rather I would have this handled (more) robust by BCBio it self, based on no or excessive coverage found in the BAM / BED files. 

Maybe it's also an idea to use some default or YAML supplied max coverage somewhere between 100 and 1000 X coverage?

I would also already appreciate more information about where and how the callable regions are currently calculated so I could investigate / attempt to  fix this myself. 

Best Regards,

Neill

**Comment by chapmanb** [comment]
Neill;
Sorry about the issues. The logic in bcbio that tries to find these high depth regions is here:

https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/bam/highdepth.py#L26

It should screen out any regions with more than 20x the median coverage in 250bp windows, so it seems like it should capture this.

Is this a whole genome, or capture target region? If the later, we don't try to exclude high depth regions for fear of removing regions due to capture (rather than repetitive regions). 

If that's the problem and you still want to screen you could manually set `coverage_interval: genome`:

https://bcbio-nextgen.readthedocs.org/en/latest/contents/configuration.html#coverage-information

If it's not the problem you might want to look at the `align/*-highdepth.bed` and `align/*-stats.yaml` to see if these help explain why it misses identifying the region, and we can try to debug more. Hope this helps.

**Comment by NeillGibson** [comment]
Thank you Brad for the quick reply and the information. I will investigate further. 

Another option that I am considering is trying to fix this (temporary) by setting higher input thresholds for Freebayes (--min-mapping-quality 10 or 20 , --min-supporting-mapping-quality 50)  and if these are not enough maybe  --use-best-n-alleles 10.

**Comment by NeillGibson** [comment]
Hi Brad, 

I think I have narrowed the issue considerably down. 

None of my bcbio bwa + freebayes analysis runs have a `align/*-highdepth.bed` or `align/*-stats.yaml` files. 

According to the commands log sambamba depth is not executed. 
`grep "sambamba depth" log/bcbio-nextgen-commands.log`

There is also no mention of identifying high coverage regions in the bed file.
```
grep  "Identify high coverage regions"  log/bcbio-nextgen-commands.log
grep  "high"  log/bcbio-nextgen-commands.log
```

Some runs have different types of coverage but also run were all samples have 'genome' coverage assigned don't have the expected bed and or yaml file. Or any mention of identifying high depth regions in the log file. 

I am trying to add some Python print statements to the functions that should handle the calculation of the high depth regions and the subsequent removal to see what is going on.  Should these all appear in the debug.log? Or is there a better way to debug?

An example yaml file that I am using is this one. Here I am starting from aligned bam files but I have the same issue when starting from fastq files. 
```
# Template for whole genome Illumina variant calling with FreeBayes
# This is a GATK-free pipeline without post-alignment BAM pre-processing
# (recalibration and realignment)
---
details:
  - analysis: variant2
    genome_build: MY_REFERENCE
    description:
    # to do multi-sample variant calling, assign samples the same metadata / batch
    metadata:
      batch: Project-123
    algorithm:
      aligner: false
      mark_duplicates: false
      recalibrate: false
      realign: false
      variantcaller: freebayes
      nomap_split_targets: 300
      effects: false
      tools_off:
      - gemini
      # for targetted projects, set the region
      # variant_regions: /path/to/your.bed
resources:
    freebayes:
        options: [--genotype-qualities, --min-mapping-quality 20   ]
```

**Comment by chapmanb** [comment]
Neill;
Sorry about the issues with this. I'm not exactly sure what is going on here as it seems like it should calculate them as long as you have `genome` coverage. I updated the development version with a bit more debugging which might help. Also I made a mistake in mentioning where the files should be, they'll be in:
```
align/sample_name/*-highdepth.bed
align/sample_name/*-highdepth-stats.yaml
```
(or `prealign/sample_name` if you're not doing alignment).

The best place to look where things are running is in `log/bcbio-nextgen-debug.log` (or on the console) where you'll see output like:
```
Processing reference #1 (chrM)
Processing reference #2 (chr22)
[2016-04-20T17:23Z] Identify high coverage regions
```
You can add print statements (to print to the console) or logger.info statements (to print to `log/bcbio-nextgen-debug.log`) when it determines whether to call high depth checks or in the run itself:

https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/pipeline/sample.py#L185
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/bam/highdepth.py#L36

Thanks for helping dig into this problem.

**Comment by NeillGibson** [comment]
Hi Brad, 

Thank you for the additional information, also on how to debug with the logger. Very helpful.

Somehow `dd.get_coverage_interval(data)` retuns None in the highdepth.py script. 

Therefore this statement equals to false because of the second condition (== "genome")   not being met. 
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/bam/highdepth.py#L36


Strange thing is that I have a log statement that mentions that the coverage is genome. But this statement is after my debug statements. 

So based on the order of the log statements the `coverage_interval` variable seems to be set after being required by the high depth script.  Maybe though this is just some async logging issue but I see it for both my two testing samples and I guess this happened for all samples in my previous larger sample sets. 

The call to assigning the coverage in `sample.py: postprocess_alignment` also seems to be after the call for calculating the high depth regions (line 191 for setting the coverage interval comes after line 185 for calculating the high depth)  :
https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/pipeline/sample.py#L191

But then I don't know why this would work for anybody else. 
```
[2016-04-20T20:06Z] noden17: assigned coverage is
[2016-04-20T20:06Z] noden17: None
[2016-04-20T20:06Z] noden17: Output bed file does not exist is
[2016-04-20T20:06Z] noden17: True
[2016-04-20T20:06Z] noden17: genome coverage ==  genome is
[2016-04-20T20:06Z] noden17: False
[2016-04-20T20:06Z] noden17: file does not exist and coverage is genome statement is
[2016-04-20T20:06Z] noden17: False
[2016-04-20T20:06Z] noden17: Resource requests: ; memory: 1.00; cores: 1
[2016-04-20T20:06Z] noden17: Configuring 10 jobs to run, using 1 cores each with 1.00g of memory reserved for each job
[2016-04-20T20:06Z] noden17: Assigned coverage as 'genome' with 88.2% genome coverage and 0.0% offtarget coverage
```

**Comment by chapmanb** [comment]
Neill;
Thanks so much. Apologies, this is my mistake. After refactoring the coverage inference got moved after high depth calculation and this is being skipped if not specified. Sorry for the issue and for completely missing this while I was starting at the code trying to debug. I pushed a fix that will resolve this if you get the latest development and will hopefully correctly identify and skip processing for the high depth regions. Sorry again about all the work to track this down and thanks for the help in identifying it.
